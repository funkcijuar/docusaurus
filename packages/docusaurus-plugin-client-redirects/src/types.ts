/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Props} from '@docusaurus/types';

export type PluginOptions = {
  fromExtensions: string[];
  toExtensions: string[];
  createRedirects?: RedirectsCreator;
};

export type UserPluginOptions = Partial<PluginOptions>;

// For a given existing route path,
// return all the paths from which we should redirect from
export type RedirectsCreator = (
  routePath: string,
) => string[] | null | undefined;

// Having an in-memory representation of wanted redirects is easier to test
export type RedirectMetadata = {
  fromRoutePath: string;
  toRoutePath: string;
  toUrl: string;
  redirectPageContent: string;
  redirectAbsoluteFilePath: string;
};

export type PluginContext = Pick<
  Props,
  'routesPaths' | 'siteConfig' | 'outDir'
> & {
  options: PluginOptions;
};
