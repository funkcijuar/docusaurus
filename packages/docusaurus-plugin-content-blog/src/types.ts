import {Loader} from 'webpack';

export interface BlogContent {
  blogPosts: BlogPost[];
  blogListPaginated: BlogPaginated[];
  blogTags: BlogTags;
  blogTagsListPath: string;
}

export interface DateLink {
  date: Date;
  link: string;
}

export interface PluginOptions {
  path: string;
  routeBasePath: string;
  include: string[];
  postsPerPage: number;
  blogListComponent: string;
  blogPostComponent: string;
  blogTagsListComponent: string;
  blogTagsPostsComponent: string;
  remarkPlugins: string[];
  rehypePlugins: string[];
  truncateMarker: RegExp | string;
}

export interface BlogTags {
  [key: string]: BlogTag;
}

export interface BlogTag {
  name: string;
  items: string[];
  permalink: string;
}

export interface BlogPost {
  id: string;
  metadata: MetaData;
}

export interface BlogPaginated {
  metadata: MetaData;
  items: string[];
}

export interface MetaData {
  permalink: string;
  source: string;
  description: string;
  date: Date;
  tags: (Tag | string)[];
  title: string;
}

export interface Tag {
  label: string;
  permalink: string;
}

export interface BlogItemsToModules {
  [key: string]: MetaDataWithPath;
}

export interface MetaDataWithPath {
  metadata: MetaData;
  metadataPath: string;
}

export interface TagsModule {
  [key: string]: TagModule;
}

export interface TagModule {
  allTagsPath: string;
  slug: string;
  name: string;
  count: number;
  permalink: string;
}

export interface ConfigureWebpackUtils {
  getStyleLoaders: (
    isServer: boolean,
    cssOptions: {
      [key: string]: any;
    },
  ) => Loader[];
  getCacheLoader: (isServer: boolean, cacheOptions?: {}) => Loader | null;
  getBabelLoader: (isServer: boolean, babelOptions?: {}) => Loader;
}
