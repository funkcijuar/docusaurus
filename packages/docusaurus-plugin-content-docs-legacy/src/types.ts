/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
export interface PluginOptions {
  path: string;
  routeBasePath: string;
  include: string[];
  sidebarPath: string;
  docLayoutComponent: string;
  docItemComponent: string;
  remarkPlugins: string[];
  rehypePlugins: string[];
}

export type SidebarItemDoc = {
  type: string;
  id: string;
};

export interface SidebarItemLink {
  type: string;
  href: string;
  label: string;
}

export interface SidebarItemCategory {
  type: string;
  label: string;
  items: SidebarItem[];
}

export interface SidebarItemCategoryRaw {
  type: string;
  label: string;
  items: SidebarItemRaw[];
}

export type SidebarItem =
  | SidebarItemDoc
  | SidebarItemLink
  | SidebarItemCategory;
export type SidebarItemRaw =
  | string
  | SidebarItemDoc
  | SidebarItemLink
  | SidebarItemCategoryRaw;

// Sidebar given by user that is not normalized yet. e.g: sidebars.json
export interface SidebarRaw {
  [sidebarId: string]: {
    [sidebarCategory: string]: SidebarItemRaw[];
  };
}

export interface Sidebar {
  [sidebarId: string]: SidebarItemCategory[];
}

export interface OrderMetadata {
  previous?: string;
  next?: string;
  sidebar?: string;
  category?: string;
  subCategory?: string;
}

export interface Order {
  [id: string]: OrderMetadata;
}

export interface MetadataRaw extends OrderMetadata {
  id: string;
  title: string;
  description: string;
  source: string;
  permalink: string;
}

export interface Metadata extends MetadataRaw {
  previous_title?: string;
  next_title?: string;
}

export interface DocsMetadata {
  [id: string]: Metadata;
}

export interface SourceToPermalink {
  [source: string]: string;
}

export interface PermalinkToId {
  [permalink: string]: string;
}

export interface LoadedContent {
  docs: DocsMetadata;
  docsDir: string;
  docsSidebars: Sidebar;
  sourceToPermalink: SourceToPermalink;
  permalinkToId: PermalinkToId;
}
