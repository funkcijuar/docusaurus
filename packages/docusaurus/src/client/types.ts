import {DocusaurusConfig} from '../server/types';

export interface DocusaurusContext {
  siteConfig?: DocusaurusConfig;
}
