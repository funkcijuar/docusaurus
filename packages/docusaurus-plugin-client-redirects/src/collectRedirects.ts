/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {flatten, uniqBy} from 'lodash';
import {
  RedirectsCreator,
  PluginContext,
  RedirectMetadata,
  PluginOptions,
} from './types';
import {
  fromExtensionsRedirectCreator,
  toExtensionsRedirectCreator,
} from './redirectCreators';

export default function collectRedirects(
  pluginContext: PluginContext,
): RedirectMetadata[] {
  const redirects = doCollectRedirects(pluginContext);
  return filterUnwantedRedirects(redirects, pluginContext);
}

function filterUnwantedRedirects(
  redirects: RedirectMetadata[],
  pluginContext: PluginContext,
): RedirectMetadata[] {
  // TODO how should we warn the user of filtered redirects?

  // we don't want to create twice the same redirect
  redirects = uniqBy(redirects, (redirect) => redirect.fromRoutePath);

  // We don't want to override an existing route
  redirects = redirects.filter(
    (redirect) => !pluginContext.routesPaths.includes(redirect.fromRoutePath),
  );

  return redirects;
}

function doCollectRedirects(pluginContext: PluginContext): RedirectMetadata[] {
  const redirectsCreators: RedirectsCreator[] = buildRedirectCreators(
    pluginContext.options,
  );

  return flatten(
    redirectsCreators.map((redirectCreator) => {
      return createRoutesPathsRedirects(redirectCreator, pluginContext);
    }),
  );
}

function buildRedirectCreators(options: PluginOptions): RedirectsCreator[] {
  const noopRedirectCreator: RedirectsCreator = (_routePath: string) => [];
  return [
    fromExtensionsRedirectCreator(options.fromExtensions),
    toExtensionsRedirectCreator(options.toExtensions),
    options.createRedirects ?? noopRedirectCreator,
  ];
}

// Create all redirects for a list of route path
function createRoutesPathsRedirects(
  redirectCreator: RedirectsCreator,
  pluginContext: PluginContext,
): RedirectMetadata[] {
  return flatten(
    pluginContext.routesPaths.map((routePath) =>
      createRoutePathRedirects(routePath, redirectCreator),
    ),
  );
}

// Create all redirects for a single route path
function createRoutePathRedirects(
  routePath: string,
  redirectCreator: RedirectsCreator,
): RedirectMetadata[] {
  const fromRoutePaths: string[] = redirectCreator(routePath) ?? [];
  return fromRoutePaths.map((fromRoutePath) => {
    return {
      fromRoutePath,
      toRoutePath: routePath,
    };
  });
}
