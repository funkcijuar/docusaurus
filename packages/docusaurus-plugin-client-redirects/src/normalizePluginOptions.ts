/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  PluginOptions,
  RedirectOption,
  RedirectsCreator,
  UserPluginOptions,
} from './types';
import * as Yup from 'yup';
import {PathnameValidator} from './redirectValidation';

export const DefaultPluginOptions: PluginOptions = {
  fromExtensions: [],
  toExtensions: [],
  redirects: [],
};

function isRedirectsCreator(value: any): value is RedirectsCreator | undefined {
  if (value === null || typeof value === 'undefined') {
    return true;
  }
  return value instanceof Function;
}

const RedirectPluginOptionValidation = Yup.object<RedirectOption>({
  to: PathnameValidator.required(),
  // wasn't able to use .when("from")...had cyclic dependency error
  // (https://stackoverflow.com/a/56866941/82609)
  from: Yup.mixed<string | string[]>().test({
    name: 'from',
    message: '${path} contains invalid redirection value',
    test: (from) => {
      return Array.isArray(from)
        ? Yup.array()
            .of(PathnameValidator.required())
            .required()
            .isValidSync(from)
        : PathnameValidator.required().isValidSync(from);
    },
  }),
});

const UserOptionsSchema = Yup.object().shape<UserPluginOptions>({
  fromExtensions: Yup.array().of(Yup.string().required().min(0)),
  toExtensions: Yup.array().of(Yup.string().required().min(0)),
  redirects: Yup.array().of(RedirectPluginOptionValidation) as any, // TODO Yup expect weird typing here
  createRedirects: Yup.mixed().test(
    'createRedirects',
    'createRedirects should be a function',
    isRedirectsCreator,
  ),
});

function validateUserOptions(userOptions: UserPluginOptions) {
  UserOptionsSchema.validateSync(userOptions, {
    abortEarly: true,
    strict: true,
  });
}

export default function normalizePluginOptions(
  userPluginOptions: UserPluginOptions = {},
): PluginOptions {
  validateUserOptions(userPluginOptions);
  return {...DefaultPluginOptions, ...userPluginOptions};
}
