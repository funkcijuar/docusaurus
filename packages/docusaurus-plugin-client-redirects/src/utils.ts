/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';

export function addTrailingSlash(str: string) {
  return str.endsWith('/') ? str : `${str}/`;
}

export function removeTrailingSlash(str: string) {
  return str.endsWith('/') ? str.slice(0, -1) : str;
}

// TODO does this function already exist?
export function getFilePathForRoutePath(routePath: string) {
  const fileName = path.basename(routePath);
  const filePath = path.dirname(routePath);
  return path.join(filePath, `${fileName}/index.html`);
}
