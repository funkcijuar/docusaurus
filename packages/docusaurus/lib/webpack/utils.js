/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const cacheLoaderVersion = require('cache-loader/package.json').version;
const merge = require('webpack-merge');

// Modify the generated webpack config with normal webpack config.
function applyConfigureWebpack(userConfig, config, isServer) {
  if (typeof userConfig === 'object') {
    return merge(config, userConfig);
  }
  if (typeof userConfig === 'function') {
    const res = userConfig(config, isServer);
    if (res && typeof res === 'object') {
      return merge(config, res);
    }
  }
  return config;
}

// Utility method to get style loaders
function getStyleLoaders(isServer, cssOptions) {
  const isProd = process.env.NODE_ENV === 'production';
  const loaders = [
    !isServer &&
      isProd && {
        loader: MiniCssExtractPlugin.loader,
      },
    !isServer && !isProd && require.resolve('style-loader'),
    {
      loader: isServer
        ? require.resolve('css-loader/locals')
        : require.resolve('css-loader'),
      options: cssOptions,
    },
  ].filter(Boolean);
  return loaders;
}

function getCacheLoader(isServer, cacheOptions) {
  return {
    loader: require.resolve('cache-loader'),
    options: Object.assign(
      {
        cacheIdentifier: `cache-loader:${cacheLoaderVersion}${isServer}`,
      },
      cacheOptions,
    ),
  };
}

function getBabelLoader(isServer, babelOptions) {
  return {
    loader: require.resolve('babel-loader'),
    options: Object.assign(
      {
        babelrc: false,
        configFile: false,
        presets: ['@babel/env', '@babel/react'],
        plugins: [
          'react-hot-loader/babel',
          isServer ? 'dynamic-import-node' : '@babel/syntax-dynamic-import',
          'react-loadable/babel',
        ],
      },
      babelOptions,
    ),
  };
}

module.exports = {
  getBabelLoader,
  getCacheLoader,
  getStyleLoaders,
  applyConfigureWebpack,
};
