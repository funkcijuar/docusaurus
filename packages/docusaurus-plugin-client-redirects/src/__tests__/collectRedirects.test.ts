/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {PluginContext, UserPluginOptions} from '../types';
import collectRedirects from '../collectRedirects';
import normalizePluginOptions from '../normalizePluginOptions';
import {removeTrailingSlash} from '../utils';

function createTestPluginContext(
  options?: UserPluginOptions,
  routesPaths: string[] = [],
): PluginContext {
  return {
    outDir: '/tmp',
    baseUrl: 'https://docusaurus.io',
    routesPaths: routesPaths,
    options: normalizePluginOptions(options),
  };
}

describe('collectRedirects', () => {
  test('should collect no redirect for undefined config', () => {
    expect(
      collectRedirects(createTestPluginContext(undefined, ['/', '/path'])),
    ).toEqual([]);
  });

  test('should collect no redirect for empty config', () => {
    expect(collectRedirects(createTestPluginContext({}))).toEqual([]);
  });

  test('should collect redirects to html/exe extension', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            fromExtensions: ['html', 'exe'],
          },
          ['/', '/somePath', '/otherPath.html'],
        ),
      ),
    ).toEqual([
      {
        fromRoutePath: '/somePath.html',
        toRoutePath: '/somePath',
      },
      {
        fromRoutePath: '/somePath.exe',
        toRoutePath: '/somePath',
      },
    ]);
  });

  test('should collect redirects to html/exe extension', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            toExtensions: ['html', 'exe'],
          },
          ['/', '/somePath', '/otherPath.html'],
        ),
      ),
    ).toEqual([
      {
        fromRoutePath: '/otherPath',
        toRoutePath: '/otherPath.html',
      },
    ]);
  });

  test('should collect redirects from plugin option redirects', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            redirects: [
              {
                from: '/someLegacyPath',
                to: '/somePath',
              },
              {
                from: ['/someLegacyPathArray1', '/someLegacyPathArray2'],
                to: '/',
              },
            ],
          },
          ['/', '/somePath'],
        ),
      ),
    ).toEqual([
      {
        fromRoutePath: '/someLegacyPath',
        toRoutePath: '/somePath',
      },
      {
        fromRoutePath: '/someLegacyPathArray1',
        toRoutePath: '/',
      },
      {
        fromRoutePath: '/someLegacyPathArray2',
        toRoutePath: '/',
      },
    ]);
  });

  test('should throw if plugin option redirects contain invalid to paths', () => {
    expect(() =>
      collectRedirects(
        createTestPluginContext(
          {
            redirects: [
              {
                from: '/someLegacyPath',
                to: '/',
              },
              {
                from: '/someLegacyPath',
                to: '/this/path/does/not/exist2',
              },
              {
                from: '/someLegacyPath',
                to: '/this/path/does/not/exist2',
              },
            ],
          },
          ['/', '/someExistingPath', '/anotherExistingPath'],
        ),
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  test('should collect redirects with custom redirect creator', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            createRedirects: (routePath) => {
              return [
                `${removeTrailingSlash(routePath)}/some/path/suffix1`,
                `${removeTrailingSlash(routePath)}/some/other/path/suffix2`,
              ];
            },
          },
          ['/', '/testpath', '/otherPath.html'],
        ),
      ),
    ).toEqual([
      {
        fromRoutePath: '/some/path/suffix1',
        toRoutePath: '/',
      },
      {
        fromRoutePath: '/some/other/path/suffix2',
        toRoutePath: '/',
      },

      {
        fromRoutePath: '/testpath/some/path/suffix1',
        toRoutePath: '/testpath',
      },
      {
        fromRoutePath: '/testpath/some/other/path/suffix2',
        toRoutePath: '/testpath',
      },

      {
        fromRoutePath: '/otherPath.html/some/path/suffix1',
        toRoutePath: '/otherPath.html',
      },
      {
        fromRoutePath: '/otherPath.html/some/other/path/suffix2',
        toRoutePath: '/otherPath.html',
      },
    ]);
  });

  test('should throw if redirect creator creates invalid redirects', () => {
    expect(() =>
      collectRedirects(
        createTestPluginContext(
          {
            createRedirects: (routePath) => {
              if (routePath === '/') {
                return [
                  `https://google.com/`,
                  `//abc`,
                  `/def?queryString=toto`,
                ];
              }
              return;
            },
          },
          ['/'],
        ),
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  test('should filter unwanted redirects', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            fromExtensions: ['html', 'exe'],
            toExtensions: ['html', 'exe'],
          },
          [
            '/',
            '/somePath',
            '/somePath.html',
            '/somePath.exe',
            '/fromShouldWork.html',
            '/toShouldWork',
          ],
        ),
      ),
    ).toEqual([
      {
        fromRoutePath: '/toShouldWork.html',
        toRoutePath: '/toShouldWork',
      },
      {
        fromRoutePath: '/toShouldWork.exe',
        toRoutePath: '/toShouldWork',
      },
      {
        fromRoutePath: '/fromShouldWork',
        toRoutePath: '/fromShouldWork.html',
      },
    ]);
  });
});
