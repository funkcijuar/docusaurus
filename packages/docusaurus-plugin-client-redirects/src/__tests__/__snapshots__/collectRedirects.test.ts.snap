// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectRedirects should throw if plugin option redirects contain invalid to paths 1`] = `
"You are trying to create client-side redirections to paths that do not exist:
- /this/path/does/not/exist2
- /this/path/does/not/exist2

Valid paths you can redirect to:
- /
- /someExistingPath
- /anotherExistingPath
"
`;

exports[`collectRedirects should throw if redirect creator creates array of array redirect 1`] = `
"Some created redirects are invalid:
- {\\"fromRoutePath\\":[\\"/fromPath\\"],\\"toRoutePath\\":\\"/\\"} => Validation error: fromRoutePath must be a \`string\` type, but the final value was: \`[
  \\"\\\\\\"/fromPath\\\\\\"\\"
]\`.
"
`;

exports[`collectRedirects should throw if redirect creator creates invalid redirects 1`] = `
"Some created redirects are invalid:
- {\\"fromRoutePath\\":\\"https://google.com/\\",\\"toRoutePath\\":\\"/\\"} => Validation error: fromRoutePath is not a valid pathname. Pathname should start with / and not contain any domain or query string
- {\\"fromRoutePath\\":\\"//abc\\",\\"toRoutePath\\":\\"/\\"} => Validation error: fromRoutePath is not a valid pathname. Pathname should start with / and not contain any domain or query string
- {\\"fromRoutePath\\":\\"/def?queryString=toto\\",\\"toRoutePath\\":\\"/\\"} => Validation error: fromRoutePath is not a valid pathname. Pathname should start with / and not contain any domain or query string
"
`;
